src/NuGet.Services.Metadata.Catalog.Monitoring/Status/PackageMonitoringStatusService.cs

```diff
@@ -70,60 +97,68 @@ public PackageMonitoringStatusService(IStorageFactory storageFactory, ILogger<Pa
 
             var storage = GetStorage(state);
 
-            var statusTasks = 
+            var statusTasks =
                 (await storage.List(token))
-                .Select(listItem =>
-                    Task.Run(async () =>
-                    {
-                        return await GetPackageAsync(storage, listItem.Uri, token);
-                    })
-                );
+                .Select(listItem => GetPackageAsync(storage, listItem.Uri, token))
+                .ToList();
 
-            return 
+            return
                 (await Task.WhenAll(statusTasks))
                 .Where(s => s != null);
         }
-        
+
         public async Task UpdateAsync(PackageMonitoringStatus status, CancellationToken token)
         {
-            // Guarantee that we never have the same package in multiple states by deleting it first.
-            await DeleteAsync(status.Package, token);
+            foreach (int stateInt in Enum.GetValues(typeof(PackageState)))
+            {
+                var state = (PackageState)stateInt;
+                if (state == status.State)
+                {
+                    // Save the new status.
+                    await SaveAsync(status, token);
```

Comment:
SaveAsync and DeleteAsync have different signature. Any reason for this?

Reply:
To clarify, this service stores status in two separate locations based on whether or not validations succeeded or failed. When a package is being moved from one status to the other, it needs to delete the package from the other location.

`SaveAsync` requires the exact `PackageMonitoringStatus` that is to be saved.

`DeleteAsync` just needs the package and version that it's deleting and the status folder that it's deleting from.

I could rewrite these functions to take in the same parameters, but I don't see a benefit.